// redux-subspace/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _subspace = require('./store/subspace');

Object.defineProperty(exports, 'subspace', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_subspace).default;
  }
});

var _applyMiddleware = require('./store/applyMiddleware');

Object.defineProperty(exports, 'applyMiddleware', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_applyMiddleware).default;
  }
});

var _namespaced = require('./reducers/namespaced');

Object.defineProperty(exports, 'namespaced', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_namespaced).default;
  }
});

var _namespacedAction = require('./actions/namespacedAction');

Object.defineProperty(exports, 'namespacedAction', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_namespacedAction).default;
  }
});

var _globalAction = require('./actions/globalAction');

Object.defineProperty(exports, 'globalAction', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_globalAction).default;
  }
});

var _applyToRoot = require('./middleware/applyToRoot');

Object.defineProperty(exports, 'applyToRoot', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_applyToRoot).default;
  }
});

var _applyToNamespaceRoots = require('./middleware/applyToNamespaceRoots');

Object.defineProperty(exports, 'applyToNamespaceRoots', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_applyToNamespaceRoots).default;
  }
});

var _applyToChildren = require('./middleware/applyToChildren');

Object.defineProperty(exports, 'applyToChildren', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_applyToChildren).default;
  }
});

var _globalActions = require('./middleware/globalActions');

Object.defineProperty(exports, 'globalActions', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_globalActions).default;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
